// puer/lib/index.js
// Generated by CoffeeScript 1.4.0
(function() {
  var connectPuer, cwd, express, helper, http, path, processOptions, puer,
    __hasProp = {}.hasOwnProperty;

  http = require('http');

  path = require('path');

  express = require('express');

  connectPuer = require('./connect-puer');

  helper = require("./helper");

  cwd = process.cwd();

  processOptions = function(options) {
    return helper.merge(options, {
      port: 8000,
      reload: true,
      dir: cwd,
      launch: true,
      addon: null,
      ignored: /(\/|^)\..*|node_modules/,
      interval: 600,
      manual: false
    });
  };

  puer = module.exports = function(options) {
    var addon, app, key, server, _ref;
    if (options == null) {
      options = {};
    }
    processOptions(options);
    options.dir = path.resolve(cwd, options.dir);
    app = express();
    if (options.reload) {
      app.use(connectPuer(app, options));
    }
    _ref = helper.requireFolder(path.join(__dirname, "./addons"));
    for (key in _ref) {
      if (!__hasProp.call(_ref, key)) continue;
      addon = _ref[key];
      addon(app, options);
    }
    try {
      if (options.addon) {
        require(path.resolve(__dirname, options.addon))(app, options);
      }
    } catch (err) {

    }
    app.use(express["static"](options.dir));
    return server = (http.createServer(app)).listen(options.port, function(err) {
      if (err) {
        return helper.log("port confict, please change port use -h to show the help", "err");
      }
      helper.log("server start at localhost:" + options.port);
      if (options.launch) {
        return helper.openBrowser("http://localhost:" + options.port, function(err) {
          return helper.log(err || "puer will lauch your default browser");
        });
      }
    });
  };

  puer.connect = connectPuer;

}).call(this);
